File: swim-js/swim-runtime-js/swim-core-js/@swim/dataflow/src/main/operator/OrOutlet.ts
// Copyright 2015-2021 Swim Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Value} from "@swim/structure";
import {Inlet, AbstractOutlet, OutletInlet} from "@swim/streamlet";

/** @internal */
export class OrOutlet extends AbstractOutlet<Value> {
  constructor() {
    super();
    this.operand1Inlet = new OutletInlet<Value>(this);
    this.operand2Inlet = new OutletInlet<Value>(this);
  }

  readonly operand1Inlet: Inlet<Value>;

  readonly operand2Inlet: Inlet<Value>;

  override get(): Value {
    const operand1Input = this.operand1Inlet.input;
    const argument1 = operand1Input !== null ? operand1Input.get() : void 0;
    if (argument1 !== void 0 && argument1.isDefinite()) {
      return argument1;
    }
    const operand2Input = this.operand2Inlet.input;
    const argument2 = operand2Input !== null ? operand2Input.get() : void 0;
    if (argument2 !== void 0) {
      return argument2;
    }
    return Value.absent();
  }
}

File: swim-java/swim-runtime/swim-core/swim.dataflow/src/main/java/swim/dataflow/operator/OrOutlet.java
// Copyright 2015-2023 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.dataflow.operator;

import swim.streamlet.AbstractOutlet;
import swim.streamlet.Inlet;
import swim.streamlet.Outlet;
import swim.streamlet.OutletInlet;
import swim.structure.Value;

public final class OrOutlet extends AbstractOutlet<Value> {

  final Inlet<Value> lhsInlet;
  final Inlet<Value> rhsInlet;

  public OrOutlet() {
    this.lhsInlet = new OutletInlet<Value>(this);
    this.rhsInlet = new OutletInlet<Value>(this);
  }

  public Inlet<Value> lhsInlet() {
    return this.lhsInlet;
  }

  public Inlet<Value> rhsInlet() {
    return this.rhsInlet;
  }

  @Override
  public Value get() {
    final Outlet<? extends Value> lhsInput = this.lhsInlet.input();
    final Value lhs = lhsInput != null ? lhsInput.get() : null;
    if (lhs != null && lhs.isDefinite()) {
      return lhs;
    }
    final Outlet<? extends Value> rhsInput = this.rhsInlet.input();
    final Value rhs = rhsInput != null ? rhsInput.get() : null;
    if (rhs != null) {
      return rhs;
    }
    return Value.absent();
  }

}
